src/main/java/reactor/ipc/netty/http/server/ConnectionInfo.java

```diff
@@ -0,0 +1,146 @@
+import reactor.ipc.netty.options.InetSocketAddressUtil;
+
+/**
+ * Resolve information about the current connection, including the
+ * host (server) address, the remote (client) address and the scheme.
+ *
+ * <p>Depending on the chosen factory method, the information
+ * can be retrieved directly from the channel or additionally
+ * using the {@code "Forwarded"}, or {@code "X-Forwarded-*"}
+ * HTTP request headers.
+ *
+ * @author Brian Clozel
+ * @since 0.8
+ * @see <a href="https://tools.ietf.org/html/rfc7239">rfc7239</a>
+ */
+public final class ConnectionInfo {
+
+	private static final String FORWARDED_HEADER = "Forwarded";
+	private static final Pattern FORWARDED_HOST_PATTERN = Pattern.compile("host=\"?([^;,\"]+)\"?");
+	private static final Pattern FORWARDED_PROTO_PATTERN = Pattern.compile("proto=\"?([^;,\"]+)\"?");
+	private static final Pattern FORWARDED_FOR_PATTERN = Pattern.compile("for=\"?([^;,\"]+)\"?");
+
+	private static final String XFORWARDED_IP_HEADER = "X-Forwarded-For";
+	private static final String XFORWARDED_HOST_HEADER = "X-Forwarded-Host";
+	private static final String XFORWARDED_PORT_HEADER = "X-Forwarded-Port";
+	private static final String XFORWARDED_PROTO_HEADER = "X-Forwarded-Proto";
+
+	private final InetSocketAddress hostAddress;
+
+	private final InetSocketAddress remoteAddress;
+
+	private final String scheme;
+
+	/**
+	 * Retrieve the connection information from the current connection directly
+	 * @param request the current server request
+	 * @param channel the current channel
+	 * @return the connection information
+	 */
+	public static ConnectionInfo newConnectionInfo(HttpServerRequest request, SocketChannel channel) {
+		InetSocketAddress hostAddress = channel.localAddress();
+		InetSocketAddress remoteAddress = channel.remoteAddress();
+		String scheme = channel.pipeline().get(SslHandler.class) != null ? "https" : "http";
+		return new ConnectionInfo(hostAddress, remoteAddress, scheme);
+	}
+
+	/**
+	 * Retrieve the connection information from the {@code "Forwarded"}/{@code "X-Forwarded-*"}
+	 * HTTP request headers, or from the current connection directly if none are found.
+	 * @param request the current server request
+	 * @param channel the current channel
+	 * @return the connection information
+	 */
+	public static ConnectionInfo newForwardedConnectionInfo(HttpServerRequest request, SocketChannel channel) {
+		if (request.requestHeaders().contains(FORWARDED_HEADER)) {
+			return parseForwardedInfo(request, channel);
+		}
+		else {
+			return parseXForwardedInfo(request, channel);
+		}
+	}
+
+	private static ConnectionInfo parseForwardedInfo(HttpServerRequest request, SocketChannel channel) {
+		InetSocketAddress hostAddress = channel.localAddress();
+		InetSocketAddress remoteAddress = channel.remoteAddress();
+		String scheme = channel.pipeline().get(SslHandler.class) != null ? "https" : "http";
+
+		String forwarded = request.requestHeaders().get(FORWARDED_HEADER).split(",")[0];
```

Comment:
is splitting on the comma `,` character ok? quoted strings in headers being allowed to contain commas BUT in this particular case:
 - `proto=` definition is same as URI scheme (no commas allowed 👍 )
 - `for=` definition is that of a Node identifier (no commas allowed 👍 )
 - `host=` definition is that of the Host header field / HTTP URI. This one is a bit more complex, but I don't think unencoded commas can appear (🤔)

So this should be ok, might want to dig into the `host` one.

Reply:
https://tools.ietf.org/html/rfc7230#section-3.2.6

`,` is part of the US-ASCII delimiters, so not allowed there AFAIU.

Reply:
that's for the `token` part (here `host=`), but the value can be a quoted string, which may contain characters in the `%x23-5B` range (including the comma `%x2C`), in addition to `%x21` (`!`) but not `%x22` (double quotes)

Reply:
I don't get it - isn't that encoded still at that point?
I just tried sending a `Forward: for="exa%x2Cmple.com"` header and `serverRequest.remoteAddress().getHostString()` does return `"exa%x2Cmple.com"`. Do you have a test case in mind that reproduces the issue?

Reply:
nope, just had a gut feeling this might be "too simple". all that is my limited understanding of the RFC, which I don't deal with as often as you guys ;) if that's working well enough for SPR, then that's probably good 👍 just challenging the assumptions a bit.

from what I understand of the ABNF notation used in the RFC, `%x2C` means "a literal comma", not "a comma encoded as %x2C" [1] (so that might be were I'm interpreting things wrong).

[1] see [ABNF rfc section 3.4](https://tools.ietf.org/html/rfc5234#section-3.4)

Reply:
You may be right, actually.
At the same time it seems commas aren't allowed in hostnames (in other RFCs) and it looks like Tomcat is doing the same: https://github.com/apache/tomcat/blob/trunk/java/org/apache/catalina/valves/RemoteIpValve.java#L354